transaction/core/src/ring_ct/rct_bulletproofs.rs

```diff
@@ -462,10 +501,68 @@ fn sign_with_balance_check<CSPRNG: RngCore + CryptoRng, S: RingSigner + ?Sized>(
     signer: &S,
     rng: &mut CSPRNG,
 ) -> Result<SignatureRctBulletproofs, Error> {
+    let ViewOnlySigningData {
+        extended_message: extended_message_digest,
+        pseudo_output_blindings,
+        pseudo_output_commitments,
+        range_proofs,
+        range_proof_bytes,
+        pseudo_output_token_ids,
+        output_token_ids,
+        ..
+    } = get_view_only_signing_data(
+        block_version,
+        message,
+        rings,
+        output_secrets,
+        fee,
+        check_value_is_preserved,
+        rng,
+    )
+    .unwrap();
+
+    // Prove that the signer is allowed to spend a public key in each ring, and that
+    // the input's value equals the value of the pseudo_output.
+    let ring_signatures: Vec<RingMLSAG> = rings
+        .iter()
+        .zip(pseudo_output_blindings)
+        .map(
+            |(ring, pseudo_output_blinding)| -> Result<RingMLSAG, Error> {
+                Ok(match ring {
+                    InputRing::ViewOnly(_) => {
+                        return Err(Error::ViewOnlyRingCannotBeSigned);
+                    }
+                    InputRing::Signable(ring) => {
+                        signer.sign(&extended_message_digest, ring, pseudo_output_blinding, rng)?
+                    }
+                    InputRing::Presigned(ring) => ring.mlsag.clone(),
+                })
+            },
+        )
+        .collect::<Result<_, _>>()?;
+
+    Ok(SignatureRctBulletproofs {
+        ring_signatures,
+        pseudo_output_commitments,
+        range_proof_bytes,
+        range_proofs,
+        pseudo_output_token_ids,
+        output_token_ids,
+    })
+}
+
+fn get_view_only_signing_data<CSPRNG: RngCore + CryptoRng>(
```

Comment:
Even if internal to this module, it should have documentation explaining what it does and what are the arguments. This should be mostly similar to `sign_with_balance_check`